proposals/crd-support.md

```diff
@@ -0,0 +1,52 @@
+# Better support for CRDs
+
+## Summary
+This is an outline for how CRDs can be handled in Helm. There are at least two ways to do this currently, but both have issues. With the implementation suggested here, CRDs are getting special treatment to make sure that they are installed correctly and are updated and deleted in a safe way. CRDs will be managed as part of a release, but will be orphaned rather than deleted when a release is removed. Similar, CRDs will be adopted into a release if they already exist.
+
+## Motivation
+The current ways to handle CRDs in Helm have issues and these are known to have caused real problems for users (more details in [#4863](https://github.com/helm/helm/issues/4863)). The goal here is to handle CRDs in a way that is intuitive and flexible for chart developers, as well as easy to use and safe for helm users. The goal is that this could be introduced in Helm 2 as well as make it into Helm 3.
+
+## Goals
+* The solution should not be dependent on versioned CRDs. Even though these are in development, a solution needs to work with existing CRDs. But we would like a solution that can work with/take advantage of versioned CRDs when they become available.
+* It should support existing installed charts that use CRDs, regardless of which of the existing ways to install CRDs was used.
+* CRDs should never be deleted unless the user has explicitely asked for it. Since CRDs are global in a Kubernetes cluster, we can never know for sure that it is no longer in use.
+* CRDs should never be updated unless the user has explicitely asked for it. Updating CRDs could render them unusable for some of the services/applications that depend on them.
+* Mixing CRDs and CRs of the type in the same chart should be supported.
+
+
+## Implementation
+This proposal takes a very conservative approach on any changes to CRDs, but the hope is that it will not be to inconvenient. If CRDs are either only used by a single chart, or is used by many charts but rarely updated (like App CRD), updates should usually be possible without too much inconvenience for the user.
+
+CRDs need to be handled correctly through all the phases of a Helm release:
+
+### Install
+* CRDs should be the first resource processed during installation. CRDs must be installed into the cluster before any CRs belonging to the CRD.
+* If a chart contains one or more CRDs, the helm install process should for each of them, check if the CRD is already installed. 
+  * If it is, Helm should check if the installed CRD is identical to the one contained in the chart. 
```

Comment:
I'm working with the install case and attempting to figure out the best way to determine if something is identical. Ideally, we can do idempotent installs. To do a comparison between what we have from the chart and what the server knows I'm noticing we would need to:

- drop the status from the response from the server as well as some fields (e.g., `metadata.uuid`)
- Walk the whole object and do a comparison because ordering is different so you can't hash

There are still issues because:

- Something may add a field that is ok for an end user to also add. For example, in a test I just ran something added to the `spec.additionalPrinterColumns` that was not in the local copy (no mutating webhooks installed either). Since this is in beta there is a good chance these changes may continue to happen. From experience we know that some changes continue to happen even after GA.
- How do we handle mutating webhooks that can alter objects when they come in
- Some tools apply their own annotations. Installing cert-manater tells you to use kubectl to add the CRDs. kubectl in tern adds some configuration (`last-applied-configuration`) as an annotation

How can we identify if the CRD has changed with so many outside sources having touched it? I'm open to suggestions.